metadata/reportssql/drug_functions.sql

```diff
@@ -276,7 +276,135 @@ BEGIN
     RETURN (result); 
 END$$ 
 
-DELIMITER ; 
+DELIMITER ;
+
+-- retrieveRegimenSwitchARVandDate
+
+DROP PROCEDURE IF EXISTS retrieveRegimenSwitchARVandDate;
+
+DELIMITER $$
+CREATE PROCEDURE retrieveRegimenSwitchARVandDate(
+    IN p_patientId INT(11),
+    IN p_startDate DATE,
+    IN p_endDate DATE,
+    OUT p_regimen VARCHAR(250),
+    OUT p_switchDate DATE
+    )
+    DETERMINISTIC
+    proc_retrieve_regimen_switch_and_date:BEGIN
+
+    DECLARE currentEncounterId INT(11);
+    DECLARE previousEncounterId INT(11);
+    DECLARE currentRegimen VARCHAR(250);
+    DECLARE previousRegimen VARCHAR(250);
+    DECLARE switchDate DATE;
+
+    SELECT o.encounter_id INTO currentEncounterId
+    FROM orders o
+    JOIN drug_order do ON do.order_id = o.order_id
+    JOIN drug d ON d.drug_id = do.drug_inventory_id AND d.retired = 0
+    WHERE o.patient_id = p_patientId AND o.voided = 0
+        AND o.date_created BETWEEN p_startDate AND p_endDate
+        AND drugIsARV(d.concept_id)
+        AND drugOrderIsDispensed(o.patient_id, o.order_id)
+    ORDER BY o.date_created DESC
+    LIMIT 1;
+
+    IF (currentEncounterId IS NULL) THEN
+        LEAVE proc_retrieve_regimen_switch_and_date;
+    END IF;
+
+    SELECT o.encounter_id INTO previousEncounterId
+    FROM orders o
+    JOIN drug_order do ON do.order_id = o.order_id
+    JOIN drug d ON d.drug_id = do.drug_inventory_id AND d.retired = 0
+    WHERE o.patient_id = p_patientId AND o.voided = 0
+        AND o.date_created BETWEEN p_startDate AND p_endDate
+        AND drugIsARV(d.concept_id)
+        AND drugOrderIsDispensed(o.patient_id, o.order_id)
+        AND o.encounter_id <> currentEncounterId
```

Comment:
@tmvumbi2 Isn't it better to use o.encounter_id < currentEncounterId and not check the date for this one?